cmd/state.go

```diff
@@ -0,0 +1,167 @@
+
+	var options []string
+	optionMap := make(map[string]*resource.State)
+	for _, ambiguousResource := range candidateResources {
+		// Prompt the user to select from a list of IDs, since these resources are known to all have the same URN.
+		message := fmt.Sprintf("Resource %q", ambiguousResource.ID)
+		if ambiguousResource.Protect {
+			message += " (Protected)"
+		}
+
+		if ambiguousResource.Delete {
+			message += " (Pending Deletion)"
+		}
+
+		options = append(options, message)
+		optionMap[message] = ambiguousResource
+	}
+
+	var option string
+	if err := survey.AskOne(&survey.Select{
+		Message:  prompt,
+		Options:  options,
+		PageSize: len(options),
+	}, &option, nil); err != nil {
+		return nil, errors.New("no resource selected")
+	}
+
+	return optionMap[option], nil
+}
+
+// runStateEdit runs the given state edit function on a resource with the given URN in the current stack.
+func runStateEdit(urn resource.URN, operation func(snap *deploy.Snapshot, res *resource.State) error) error {
+	return runTotalStateEdit(func(opts display.Options, snap *deploy.Snapshot) error {
+		res, err := locateStackResource(opts, snap, urn)
+		if err != nil {
+			return err
+		}
+
+		if res == nil {
+			return errors.Errorf("No such resource %q exists in the current state", urn)
+		}
+
+		return operation(snap, res)
+	})
+}
+
+// runTotalStateEdit runs a snapshot-mutating function on the entirity of the current stack's snapshot. Before mutating
+// the snapshot, the user is prompted for confirmation if the current session is interactive.
+func runTotalStateEdit(operation func(opts display.Options, snap *deploy.Snapshot) error) error {
+	opts := display.Options{
+		Color: cmdutil.GetGlobalColorization(),
+	}
+	s, err := requireCurrentStack(true, opts, true /*setCurrent*/)
+	if err != nil {
+		return err
+	}
+	snap, err := s.Snapshot(commandContext())
+	if err != nil {
+		return err
+	}
+
+	if cmdutil.Interactive() {
+		confirm := false
+		surveycore.DisableColor = true
+		surveycore.QuestionIcon = ""
+		surveycore.SelectFocusIcon = opts.Color.Colorize(colors.BrightGreen + ">" + colors.Reset)
+		if err = survey.AskOne(&survey.Confirm{
+			Message: "This command will edit your stack's state directly. Confirm?",
+		}, &confirm, nil); err != nil || !confirm {
+			return errors.New("confirmation declined")
+		}
+	}
+
+	stackIsAlreadyHosed := snap.VerifyIntegrity() != nil
+	if err = operation(opts, snap); err != nil {
+		return err
+	}
+
+	// If the stack is already broken, don't bother verifying the integrity here.
+	if !stackIsAlreadyHosed {
+		contract.AssertNoErrorf(snap.VerifyIntegrity(), "state edit produced an invalid snapshot")
+	}
+	bytes, err := json.Marshal(stack.SerializeDeployment(snap))
+	if err != nil {
+		return err
+	}
+	dep := apitype.UntypedDeployment{
+		Version:    apitype.DeploymentSchemaVersionCurrent,
+		Deployment: bytes,
+	}
+	return s.ImportDeployment(commandContext(), &dep)
```

Comment:
i don't really understand the previous 15 or so lines.

Reply:
I'll add some comments.